Src/Support/Google.Apis.Core/Http/HttpClientFromMessageHandlerFactory.cs

```diff
@@ -0,0 +1,189 @@
+
+        /// <summary>
+        /// Factory for obtaining the underlying <see cref="HttpMessageHandler"/>
+        /// of the <see cref="ConfigurableHttpClient"/> returned by
+        /// <see cref="IHttpClientFactory.CreateHttpClient(CreateHttpClientArgs)"/>.
+        /// Won't be null.
+        /// </summary>
+        private readonly Func<HttpMessageHandlerOptions, ConfiguredHttpMessageHandler> _httpMessageHandlerFactory;
+
+        /// <summary>
+        /// Creates an <see cref="HttpClientFromMessageHandlerFactory"/> that will use
+        /// the given <see cref="HttpMessageHandler"/> factory for creating the inner
+        /// message handlers that will be used when creating the <see cref="ConfigurableHttpClient"/>.
+        /// </summary>
+        public HttpClientFromMessageHandlerFactory(Func<HttpMessageHandlerOptions, ConfiguredHttpMessageHandler> httpMessageHandlerFactory) =>
+            _httpMessageHandlerFactory = httpMessageHandlerFactory.ThrowIfNull(nameof(httpMessageHandlerFactory));
+
+        /// <inheritdoc/>
+        public ConfigurableHttpClient CreateHttpClient(CreateHttpClientArgs args)
+        {
+            var handler = CreateHandler(args);
+            var configurableHandler = new ConfigurableMessageHandler(handler)
+            {
+                ApplicationName = args.ApplicationName,
+                GoogleApiClientHeader = args.GoogleApiClientHeader
+            };
+
+            // We always set not to dispose the inner handler, it's not created
+            // by us so we don't control it's lifetime.
+            var client = new ConfigurableHttpClient(configurableHandler, false);
+            foreach (var initializer in args.Initializers)
+            {
+                initializer.Initialize(client);
+            }
+
+            return client;
+        }
+
+        private HttpMessageHandler CreateHandler(CreateHttpClientArgs args)
+        {
+            // We need to handle three situations in order to intercept uncompressed data where necessary
+            // while using the built-in decompression where possible.
+            // - No compression requested
+            // - Compression requested but not supported by the injected message handler
+            //   (easy; just GzipDeflateHandler on top of an interceptor on top of HttpMessagetHandler)
+            // - Compression requested and supported by the injected message handler
+            //   (complex: create two different handlers and decide which to use on a per-request basis)
+
+            // First request a message handler that may perform decompression.
+            var configuredMessageHandler = AllowDecompressionOptions.CheckItMatches(_httpMessageHandlerFactory(AllowDecompressionOptions));
+
+            if (!args.GZipEnabled)
+            {
+                // Simple: nothing will be decompressing content, so we can just intercept the original handler.
```

Comment:
This comment is a bit confusing to me (and I probably wrote it in the first place!). Do we believe that no decompression is involved because we won't be sending Accept-Encoding: gzip as a request header?

Reply:
In this case we won't actively be setting the encoding header, it's set by the `GzipDeflateHandler`. But this code as it is will now result in a bug (so thanks for catching). In this case we also need to make sure to request a handler that does no decompression (and does not potentially set the header itself) because we have been told not to through this parameter beeing false.
(It's not a problem in the current implementation because we manually configure the handler to AutomaticDecompression.None before possibly configuring it to something else)